waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+  popFollowing
+  -- Run second (or throw error if first failed)
+  case mbError of
+    Nothing -> doParse second
+    Just err -> throwError err
+
+-- | Parse "Alternative" rule. If the left parser can consume the next token in
+-- the input, run that parser. If the right parser can consume the next token,
+-- run that parser. Otherwise, report an error.
+parseAlternative :: GrammarRule -> GrammarRule -> ParserM ()
+parseAlternative left right = do
+  consumeWhitespace
+  let leftFirstTokens = findFirstTokens left
+  let rightFirstTokens = findFirstTokens right
+  nextToken <- peek
+  let nextKind = tokenKind <$> nextToken
+  if nextKind `willSucceedIn` left
+    then doParse left
+    else
+      if nextKind `willSucceedIn` right
+        then doParse right
+        else makeError nextToken (leftFirstTokens `TokenSet.union` rightFirstTokens)
+
+-- | Advance past whitespace, making nodes for the whitespace as it goes
+consumeWhitespace :: ParserM ()
+consumeWhitespace =
+  peek >>= \case
+    Nothing -> pure ()
+    Just nextToken
+      -- Consume any whitespace
+      | tokenKindIsTrivia (tokenKind nextToken) -> do
+        advance
+        let newNode =
+              SyntaxNode
+                { snodeKind = S.Token (tokenKind nextToken),
+                  snodeWidth = tokenWidth nextToken,
+                  snodeChildren = []
+                }
+        pushNode newNode
+        consumeWhitespace
+      | otherwise -> pure ()
+
+-- | Peek the immediate next token in input (including whitespace)
+peek :: ParserM (Maybe Token)
+peek =
+  gets pstateInput >>= \case
+    [] -> return Nothing
+    tok : _ -> return (Just tok)
+
+-- | Advance past the next token in the input, updating offsets.
+advance :: ParserM ()
+advance =
+  gets pstateInput >>= \case
+    [] -> modify (\s -> s {pstateLastOffset = pstateNextOffset s})
+    (consumed : remaining) -> do
+      newLastOffset <- gets pstateNextOffset
+      let newNextOffset = newLastOffset + tokenWidth consumed
+      modify (\s -> s {pstateInput = remaining, pstateLastOffset = newLastOffset, pstateNextOffset = newNextOffset})
+
+-- | @makeError actual expected@ throws an error.
+--
+-- Assumes @actual@ has not been `advance`d past yet (and does not advance past
+-- it).
+makeError :: Maybe Token -> TokenSet -> ParserM ()
+makeError Nothing eset = do
+  offset <- gets pstateNextOffset
+  throwError $ ParseError $ UnexpectedEOF offset eset
+makeError (Just actual) eset = do
+  offset <- gets pstateNextOffset
+  throwError $ ParseError $ Unexpected (Span offset (offset + tokenWidth actual)) (tokenKind actual) eset
+
+-- | Make an error node. Error node should only be produced when catching and
+-- pushing errors to the state.
+errorNode ::
+  -- | error to make an error node from
+  ParseError ->
+  -- | If "True", give the error node a width of 0. Otherwise, use the
+  -- width of the error item.
+  Bool ->
+  SyntaxNode
+errorNode (UnexpectedEOF _ _) _ =
+  SyntaxNode
```

Comment:
So you don't store actual error into the node? How is that -> you store only that it is error, and its width! Feels like a lot of information is not stored? But you don't need it?

I guess we touch here the point of duplicating errors -> you return them in a list, and you also put them in a tree. But those in the tree are missing information. Why do you even need them in the tree, if you have them in the list? On the other hand, why don't you just put them all into the tree, with full info, and then you can easily get a list if needed? Or are there some that don't have a place to be put at? What is the thinking here?